apps/prairielearn/src/middlewares/clientFingerprint.sql

```diff
@@ -0,0 +1,40 @@
+--BLOCK select_client_fingerprint
+SELECT
+  id, user_id, session_id, ip_address, user_agent, accept, accept_language
+FROM
+  client_fingerprints
+WHERE
+  user_id = $user_id
+  AND session_id = $session_id
+  AND ip_address = $ip_address
+  AND user_agent = $user_agent
+  AND accept = $accept
+  AND accept_language = $accept_language
+
+--BLOCK insert_client_fingerprint
+INSERT INTO client_fingerprints (
+  user_id,
+  session_id,
+  ip_address,
+  user_agent,
+  accept_language,
+  accept
+) 
+VALUES (
+  $user_id,
+  $session_id,
+  $ip_address,
+  $user_agent,
+  $accept_language,
+  $accept
+)
+RETURNING id
+
+--BLOCK update_assessment_instance_fingerprint
+UPDATE assessment_instances
+SET
+  last_client_fingerprint_id = $client_fingerprint_id,
+  client_fingerprint_id_change_count = $client_fingerprint_id_change_count
```

Comment:
Rather than passing in an updated count, you can instead compute this on the fly based on the previous count, which should be less susceptible to race conditions.

```suggestion
  client_fingerprint_id_change_count = client_fingerprint_id_change_count + 1
```

Reply:
This has been updated in commit b135adf.